### DISCUSSION TITLE
Questions on the Security and Privacy of Ente Auth

### DISCUSSION THREAD

**Discussion by okikio** [ente-io/ente#2772] [Q&A]
Ente Auth is a robust and well-structured end-to-end encryption system. The layered encryption approach, secure handling of keys, and thoughtful inclusion of token-based authentication are all significant strengths. The detailed recovery process also reflects a deep understanding of user needs in maintaining access while keeping their data secure. 

I do have a few questions about how the security and privacy of the encryption system works. I'd love to hear your thoughts on these questions:

### **1. Metadata Handling**

I noticed that the system encrypts user and device metadata, which is great, but even encrypted metadata can sometimes reveal patterns. Does Ente Auth use additional obfuscation techniques or minimize metadata storage to reduce this potential risk? Also, are metadata and content encrypted using separate keys to further compartmentalize potential breaches?

### **2. Security of Secondary Devices**

How is the security of secondary devices currently managed when they are added to a user's account? Are there plans to enhance this process by introducing device-specific keys or robust pairing methods like QR code-based key exchanges to prevent unauthorized access?

### **3. Revocation and Key Rotation**

In case a device is compromised, is there a plan for implementing automatic key revocation and rotation mechanisms, similar to what we've seen in other systems like Google Accounts or Microsoft Accounts? How do you handle compromised keys in the current implementation, and what challenges do you foresee with adding automatic policies for this?

### **4. Use of 2-factor and/or Passkeys**

It seems that after setting up Passkeys, 2-factor authentication is no longer available, requiring the use of a recovery key instead. Is this intended behavior, or is it something that will be adjusted in future updates?

### **5. Data Integrity Checks**

Is there a mechanism for ensuring data integrity in addition to encryption? For instance, if data were tampered with during transit, would the system detect and flag this using HMACs or digital signatures?

Thanks for taking the time and consideration. I’d love to hear your insights on these points and understand your perspective on these questions!

**Comment by marek22k** [comment]
Hi, this page is now on the first page of DuckDuckGo when you search for "ente.io encryption." Maybe the ente.io team can respond here? @mnvr @ua741 @vishnukvmd @ashilkn

**Reply by mnvr** [reply]
https://ente.io/architecture

**Reply by marek22k** [reply]
However, this only answers some of the questions to a limited extent. Am I understanding the following correctly?
1. Apart from encryption (and European data centers), no other concealment measures are taken.
2. No, the photos are only encrypted with a master key, which is encrypted by the user password.
3. No, in this case, you're out of luck. (Do you support re-encryption with a new master key?)
4. No answer found (or you would have to look at the source code).
5. Yes, this would be detected by libsodium (keyword: Poly1305 MAC).

**Reply by Damariobros** [reply]
I can answer question 4. Since the creation of this thread, Ente has implemented a dialogue box that pops up after entering your password, which asks you whether you want to use TOTP or Passkeys to sign in.

If you're using Email 2FA as well, this will come after passing Email Verification but before entering your password.

**Reply by ed-asriyan** [reply]
3. there is no option to rotate master key. user should create a new account (so the new master key is generated) and migrate all data over. https://www.reddit.com/r/enteio/comments/1q5rsj5/comment/ny4ufkr
